assets/js/googlesitekit/site/index.js

```diff
@@ -0,0 +1,210 @@
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Internal dependencies
+ */
+import { get, set } from 'googlesitekit-api';
+import {
+	FETCH_CONNECTION_INFO,
+	FETCH_RESET,
+	INITIALIZE,
+	RESET,
+	RESET_FAILURE,
+	RESET_SUCCESS,
+	RECEIVE_CONNECTION_INFO,
+	RECEIVE_CONNECTION_INFO_FAILED,
+} from './index.private';
+
+export const INITIAL_STATE = {
+	connectionInfo: null,
+	isFetchingConnectionInfo: false,
+	isResetting: false,
+};
+
+export const STORE_NAME = 'core/site';
+
+export const actions = {
+	fetchConnectionInfo() {
+		return {
+			payload: {},
+			type: FETCH_CONNECTION_INFO,
+		};
+	},
+
+	fetchReset() {
+		return {
+			payload: {},
+			type: FETCH_RESET,
+		};
+	},
+
+	initialize() {
+		return {
+			payload: {},
+			type: INITIALIZE,
+		};
+	},
+
+	receiveConnectionInfo( connectionInfo ) {
+		invariant( connectionInfo, 'connectionInfo is required.' );
+
+		return {
+			payload: { connectionInfo },
+			type: RECEIVE_CONNECTION_INFO,
+		};
+	},
+
+	receiveConnectionInfoFailed( error ) {
+		invariant( error, 'error is required.' );
+
+		return {
+			payload: { error },
+			type: RECEIVE_CONNECTION_INFO_FAILED,
+		};
+	},
+
+	*reset() {
+		try {
+			yield actions.fetchReset();
+			yield actions.resetSuccess();
+			yield actions.initialize();
+
+			return {
+				payload: {},
+				type: RESET,
+			};
+		} catch ( err ) {
+			return actions.resetFailure( err );
+		}
+	},
+
+	resetFailure( error ) {
+		invariant( error, 'error is required.' );
+		return {
+			payload: { error },
+			type: RESET_FAILURE,
+		};
+	},
+
+	resetSuccess() {
+		return {
+			payload: {},
+			type: RESET_SUCCESS,
+		};
+	},
+};
+
+export const reducer = ( state = INITIAL_STATE, action = {} ) => {
+	switch ( action.type ) {
+		case FETCH_CONNECTION_INFO: {
+			return {
+				...state,
+				isFetchingConnectionInfo: true,
+			};
+		}
+
+		case FETCH_RESET: {
+			return {
+				...state,
+				isResetting: true,
+			};
+		}
+
+		case INITIALIZE: {
+			return { ...INITIAL_STATE };
+		}
+
+		case RECEIVE_CONNECTION_INFO: {
+			const { connectionInfo } = action.payload;
+
+			return {
+				...state,
+				isFetchingConnectionInfo: false,
+				connectionInfo,
+			};
+		}
+
+		case RECEIVE_CONNECTION_INFO_FAILED: {
+			const { error } = action.payload;
+
+			return {
+				...state,
+				isFetchingConnectionInfo: false,
+				connectionInfo: { error, hasError: true },
```

Comment:
Is this an established pattern for dealing with errors? Feels rather strange to me including the error in the regular data object with a `hasError` flag. A component relying on the connection info shouldn't need to check if `hasError`, that would really clutter the codebase.

I think if this fails, it should either do so silently or (probably more appropriate) kill the application (this should never fail unless there was a severe error that we would need to fix or if there were connectivity issues).

Reply:
No, not at all. The error handling (as mentioned elsewhere in the file) is half-baked right now, but I think it's better to build half-baked versions in to test and ensure control flow/behaviour than leave them out (doing so can hide bugs).

What I've traditionally done in Redux apps is have an error handler passed to the store by components with a HOC. Something like https://github.com/mozilla/addons-frontend/blob/master/src/amo/sagas/reviews.js#L114-L117 is what I'd eventually like to have; all errors are dispatched to an errorHandler, which is usually set up like: https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/AddonReviewCard/index.js#L589 and passed to actions like so: https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/AddonReviewCard/index.js#L168-L174

Doing error handling in Redux + React can be a bit verbose but that approach has worked well for me in the past, especially as the errors can be localised to each component.

In this case though, I think you're right: checking on `hasError` throughout the codebase would be bad. This might be something appropriate to only catch in a top-level component or even throw and let our top-level `<ErrorHandler>` catch.

Okay, I'll revisit this particular one. We'll need to think about how we handle errors in the future, but I suppose this one is a niche case…

Reply:
I like those ideas. Let's revisit in the future. I think for now we can just set `connectionInfo` to its previous value or back to `null`, and maybe log a warning?